File: src/main/java/no/fintlabs/validation/OnlyOneTimeFilterTypeValidator.java
package no.fintlabs.validation;

import no.fintlabs.model.time.TimeFilter;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import java.util.Optional;
import java.util.stream.Stream;

public class OnlyOneTimeFilterTypeValidator implements ConstraintValidator<OnlyOneTimeFilterType, TimeFilter> {

    @Override
    public boolean isValid(TimeFilter timeFilter, ConstraintValidatorContext constraintValidatorContext) {
        if (timeFilter == null) {
            return true;
        }
        return Stream.of(
                        timeFilter.getOffset(),
                        timeFilter.getCurrentPeriod(),
                        timeFilter.getManual()
                )
                       .map(Optional::ofNullable)
                       .filter(Optional::isPresent)
                       .count() < 2;
    }

}

File: src/main/kotlin/no/novari/flyt/history/validation/OnlyOneTimeFilterTypeValidator.kt
package no.novari.flyt.history.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import no.novari.flyt.history.model.time.TimeFilter

class OnlyOneTimeFilterTypeValidator : ConstraintValidator<OnlyOneTimeFilterType, TimeFilter> {
    override fun isValid(
        timeFilter: TimeFilter?,
        context: ConstraintValidatorContext?,
    ): Boolean {
        if (timeFilter == null) {
            return true
        }

        return listOf(timeFilter.offset, timeFilter.currentPeriod, timeFilter.manual)
            .count { it != null } < 2
    }
}
